File: confluence_to_markdown.js
function confluenceToMarkdown() {
  // Confluenceの設定 (実際の値に置き換えてください)
  const CONFLUENCE_URL = "https://your-confluence-domain.atlassian.net";
  const SPACE_KEY = "YOUR_SPACE_KEY";
  const API_TOKEN = "YOUR_API_TOKEN";
  const OUTPUT_DIR = "markdown_files"; // Google Driveのフォルダ名
  const ROOT_PAGE_ID = "123456789";

  function getPageContent(pageId) {
    const url = `${CONFLUENCE_URL}/rest/api/content/${pageId}?expand=body.storage`;
    const headers = {
      "Authorization": `Bearer ${API_TOKEN}`,
      "Content-Type": "application/json"
    };
    const options = {
      "method": "get",
      "headers": headers,
      "muteHttpExceptions": true  // エラーをスローしない
    };
    const response = UrlFetchApp.fetch(url, options);
    if (response.getResponseCode() !== 200) {
      Logger.log(`Error fetching page ${pageId}: ${response.getContentText()}`);
      return null;
    }
    return JSON.parse(response.getContentText());
  }

  function convertHtmlToMarkdown(htmlContent) {
    // 簡易的な変換 (必要に応じてhtml2text相当の処理を追加)
    // Google Apps ScriptにはDOMParserがないため、正規表現などで簡易的に処理
    let text = htmlContent.replace(/<br\s*[\/]?>/gi, "
"); // <br> を改行に
    text = text.replace(/<[^>]*>/g, ""); // タグの除去
    text = text.replace(/&nbsp;/g, " "); // &nbsp; をスペースに
    return text;
  }

  function exportPages(rootPageId) {
    // Google Driveのフォルダを作成または取得
    let folder = DriveApp.getRootFolder().getFoldersByName(OUTPUT_DIR);
    if (folder.hasNext()) {
      folder = folder.next();
    } else {
      folder = DriveApp.getRootFolder().createFolder(OUTPUT_DIR);
    }

    // ルートページの内容を取得
    const rootData = getPageContent(rootPageId);
    if (!rootData) return; // エラー処理

    const rootTitle = rootData.title;
    const rootHtml = rootData.body.storage.value;
    const rootMarkdown = convertHtmlToMarkdown(rootHtml);

    // ルートページを保存
    const file = folder.createFile(`${rootTitle}.md`, `# ${rootTitle}

${rootMarkdown}`, MimeType.PLAIN_TEXT);
    Logger.log(`Exported: ${rootTitle}`);

    // サブページを再帰的に処理
    const childrenUrl = `${CONFLUENCE_URL}/rest/api/content?spaceKey=${SPACE_KEY}&ancestor=${rootPageId}`;
    const childrenHeaders = {
      "Authorization": `Bearer ${API_TOKEN}`
    };
    const childrenOptions = {
      "method": "get",
      "headers": childrenHeaders,
      "muteHttpExceptions": true
    };
    const childrenResponse = UrlFetchApp.fetch(childrenUrl, childrenOptions);

    if (childrenResponse.getResponseCode() !== 200) {
      Logger.log(`Error fetching children: ${childrenResponse.getContentText()}`);
      return;
    }

    const children = JSON.parse(childrenResponse.getContentText()).results || [];
    children.forEach(child => exportPages(child.id));
  }

  // スクリプトの実行
  exportPages(ROOT_PAGE_ID);
}

File: confluence_to_markdown.py
import os
import requests
from bs4 import BeautifulSoup

# Confluenceの設定
CONFLUENCE_URL = "https://your-confluence-domain.atlassian.net"
SPACE_KEY = "YOUR_SPACE_KEY"  # エクスポート対象のスペースキー
API_TOKEN = "YOUR_API_TOKEN"  # ConfluenceのAPIトークン
OUTPUT_DIR = "./markdown_files"  # 出力先ディレクトリ


def get_page_content(page_id):
    """Confluenceのページ内容を取得"""
    url = f"{CONFLUENCE_URL}/rest/api/content/{page_id}?expand=body.storage"
    headers = {
        "Authorization": f"Bearer {API_TOKEN}",
        "Content-Type": "application/json"
    }
    response = requests.get(url, headers=headers)
    return response.json()


def convert_html_to_markdown(html_content):
    """HTMLをMarkdownに変換（簡易処理）"""
    soup = BeautifulSoup(html_content, "html.parser")
    text = soup.get_text(separator="\n", strip=True)
    # さらに高度な変換が必要な場合はhtml2textライブラリを使用推奨
    return text


def export_pages(root_page_id):
    """指定されたルートページからサブページを再帰的にエクスポート"""
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    # ルートページの内容を取得
    root_data = get_page_content(root_page_id)
    root_title = root_data["title"]
    root_html = root_data["body"]["storage"]["value"]
    root_markdown = convert_html_to_markdown(root_html)

    # ルートページを保存
    with open(f"{OUTPUT_DIR}/{root_title}.md", "w") as f:
        f.write(f"# {root_title}\n\n{root_markdown}")
    print(f"Exported: {root_title}")

    # サブページを再帰的に処理
    children_url = f"{CONFLUENCE_URL}/rest/api/content?spaceKey={SPACE_KEY}&ancestor={root_page_id}"
    children_response = requests.get(
        children_url, headers={"Authorization": f"Bearer {API_TOKEN}"})
    for child in children_response.json().get("results", []):
        export_pages(child["id"])


if __name__ == "__main__":
    # ルートページのIDを指定（トップページのIDを調べる必要があります）
    ROOT_PAGE_ID = "123456789"  # 例: ConfluenceのトップページID
    export_pages(ROOT_PAGE_ID)
